### PR TITLE
Fix boolean behavior

### PR DESCRIPTION
**Pull request by akaszynski**
Current behavior of boolean cut uses ``SetOperationToIntersection``, which does not produce the desired behavior of actually cutting.  For example:

```py
>>> import pyvista
>>> cube = pyvista.Cube().triangulate().subdivide(2).clean()
>>> sphere = pyvista.Sphere(center=(0.5, 0.5, 0.5))
>>> sphere_cutting_cube = cube.boolean_cut(sphere)
>>> sphere_cutting_cube.plot(show_edges=True)
```

Produces the union of the two meshes:
wrong

---

This PR corrects this behavior to instead use ``SetOperationToDifference``, which returns the correct mesh:
right

### Fix ``boolean_difference``
Boolean difference is now ``boolean_cut``.  Needing a `intersection` operation, this PR also changes ``boolean_difference`` to ``boolean_intersection``, which ironically already has a docstring matching the behavior of `SetOperationToIntersection`.

### CODE REVIEW

File: pyvista/core/filters/poly_data.py [hunk 1]
```diff
@@ -156,7 +153,7 @@ def boolean_union(poly_data, mesh, inplace=False):
         else:
             return mesh
 
-    def boolean_difference(poly_data, mesh, inplace=False):
+    def boolean_intersection(poly_data, mesh, inplace=False):
         """Combine two meshes and retains only the volume in common between the meshes.
```
Comment (MatthewFlamm) [inline pyvista/core/filters/poly_data.py:157]: The description of the mesh parameter and return below are not correct too

**Review by MatthewFlamm — COMMENTED** [review]
I think this PR is fine as is, but a thought: It is evident that the tests do not actually test that the filters work correctly.  It is tricky however to construct the right tests.  One could test that a cut operation reduces the maximum mesh bounds in the x direction, for example, while a union operation adds to the maximum mesh bounds.

**Comment by adeak** [comment]
I never got the hang of boolean filters, because they never seem to do what I'd expect based on their set theoretical names.

I was looking forward to this PR fixing this, but I still don't get it :D I suspect this is a VTK thing, or a wrong premise on my part.

Using the new naming in this PR:

1. called "boolean add":
tmp_add
this is what I'd call a set union

2. called "boolean cut":
tmp_cut
this is what I'd call a set intersection, but I admit this is not entirely consistent because "boolean cut" is not commutative whereas set intersection is... So I'd say that `sphere - cube` is a set intersection, but `cube - sphere` looks like a set symmetric difference...

3. called "boolean intersection":
tmp_intersection
this is what I'd _also_ call a set difference, but I admit this is not entirely consistent because "boolean intersection" is commutative whereas set difference is not... So I'd say that `sphere.boolean_intersection(cube)` looks like a set difference of sphere \ cube, but so does `cube.boolean_intersection(sphere)`...

4. called "boolean union":
tmp_union
again looks like a set difference the other way around: `sphere.boolean_union(cube)` and `cube.boolean_union(sphere)` look like the set difference of cube \ sphere.


Some further surprises based on the name:
1. `cube.boolean_union(cube)` is empty. That's not what I'd expect any union to do.
2. `boolean_intersection`'s docstring says "_Combine two meshes and retains only the volume in common between the meshes._". Yet the remaining points in `cube.boolean_intersection(sphere)` all belong to the sphere, outside the cube.

**Comment by akaszynski** [comment]
> I suspect this is a VTK thing, or a wrong premise on my part.

It's both.

Honestly, VTK boolean filters aren't great, but considering that what they're trying to do is hard, it's understandable.

We clearly need more explanation on what's going on here, especially as flipping normals has a drastic effect on the behavior of these filters, and that's part of what you're seeing.  We need to expand our examples to include documentation on the effect of normals.  When both normals are outward, it seems that the direction in which the sphere "cuts" changes.

Let's rename these to "difference" rather than cut, and then proceed to document how the normals *both* the meshes affects the boolean operations.

```py
import pyvista
pyvista.set_plot_theme('document')
cube = pyvista.Cube().triangulate().subdivide(3).clean()
sphere = pyvista.Sphere(radius=0.6)
pyvista.plot([cube, sphere], cpos='xz', style='wireframe')
```
nom

```py
cube.boolean_cut(sphere).plot()
```
not_flipped


```py
sphere.flip_normals()
cube.boolean_cut(sphere).plot()
```
flipped

---
### flip cube mesh

```py
cube = pyvista.Cube().triangulate().subdivide(3).clean()
cube.flip_normals()
sphere = pyvista.Sphere(radius=0.6)
cube.boolean_cut(sphere).plot()
```
cube_not_flipped

```py
sphere.flip_normals()
cube.boolean_cut(sphere).plot()
```
cube_flipped

**Comment by akaszynski** [comment]
I think I explained it better this time around:

## Union
The intersection of two manifold meshes A and B is the mesh which is the volume of A that is also in B.
```py
import pyvista
sphere_a = pyvista.Sphere()
sphere_b = pyvista.Sphere(center=(0.5, 0, 0))
result = sphere_a.boolean_intersection(sphere_b)
pl = pyvista.Plotter()
pl.add_mesh(sphere_a, color='r', style='wireframe', line_width=3)
pl.add_mesh(sphere_b, color='b', style='wireframe', line_width=3)
pl.add_mesh(result, color='tan')
pl.camera_position = 'xz'
pl.show()
```
image

### Difference 
The difference of two manifold meshes A and B is the volume of the mesh in A not belonging to B.

```py
import pyvista
sphere_a = pyvista.Sphere()
sphere_b = pyvista.Sphere(center=(0.5, 0, 0))
result = sphere_a.boolean_difference(sphere_b)
pl = pyvista.Plotter()
pl.add_mesh(sphere_a, color='r', style='wireframe', line_width=3)
pl.add_mesh(sphere_b, color='b', style='wireframe', line_width=3)
pl.add_mesh(result, color='tan')
pl.camera_position = 'xz'
pl.show()
```
image


### Intersection
The intersection of two manifold meshes A and B is the mesh which is the volume of A that is also in B.

```py
import pyvista
sphere_a = pyvista.Sphere()
sphere_b = pyvista.Sphere(center=(0.5, 0, 0))
result = sphere_a.boolean_intersection(sphere_b)
pl = pyvista.Plotter()
pl.add_mesh(sphere_a, color='r', style='wireframe', line_width=3)
pl.add_mesh(sphere_b, color='b', style='wireframe', line_width=3)
pl.add_mesh(result, color='tan')
pl.camera_position = 'xz'
pl.show()
```
image

**Comment by adeak** [comment]
Sorry, my laptop's home partition fell apart a bit and I'm only starting to recover (all's good, but I had a few hours less free time than expected).

Thanks for the explanations. Your newest comment seems perfectly consistent. I wonder how deceptive that example is considering the "cube vs sphere" weirdnesses :smile: Also thanks for the tip about normals, that was the (/a) missing piece in my understanding. I'll see if I can get a consistent picture of these filters. Even if I don't I can review the PR.

**Comment by adeak** [comment]
I think your first code block and docstring wanted to have `union` even though it had `intersection` (like the third block).

**Comment by adeak** [comment]
OK, it's slowly starting to click. The reason why [the original example](https://dev.pyvista.org/examples/01-filter/boolean-operations.html#sphx-glr-examples-01-filter-boolean-operations-py) is weird is because the normals of that cube point inwards (perhaps it'd we worth asking why that's the case). If its normals are flipped so that they point outward (similarly to what `pv.Cube().triangulate().subdivide(5)` would do out of the box), the results start to make sense.

And I see you've updated the example to showcase this. I'm wondering about the order of (weird -> expected), but that's bikeshedding category.

I'm also wondering about backwards compatibility, having rearranged and removed filters. We should probably have `versionadded`/`versionchanged` directives for these, and a deprecation error for what's been removed. But let me do a proper review...

**Comment by akaszynski** [comment]
> And I see you've updated the example to showcase this. I'm wondering about the order of (weird -> expected), but that's bikeshedding category.

I'm going to fix the example to make it clearer.  The spheres example is probably the best showcase for this.

> I'm also wondering about backwards compatibility, having rearranged and removed filters. We should probably have versionadded/versionchanged directives for these, and a deprecation error for what's been removed. But let me do a proper review..

I'll add in some depreciation errors.  I'm jumping straight to errors since the behavior was invalid before.  I'll also add the version added/changed directives for these.

File: tests/test_polydata.py [hunk 2]
```diff
@@ -274,61 +274,72 @@ def test_edge_mask(sphere):
...
+    expected_volume = sphere.volume + sphere_shifted.volume - intersection.volume
+    assert np.isclose(union.volume, expected_volume, atol=1E-3)
 
-def test_boolean_cut_fail(plane):
-    with pytest.raises(NotAllTrianglesError):
-        plane - plane
 
+def test_boolean_intersection(sphere, sphere_shifted):
+    intersection = sphere.boolean_intersection(sphere_shifted)
+    union = sphere.boolean_union(sphere_shifted)
```
Comment (adeak) [inline tests/test_polydata.py:288]: These two are basically the same test so I think we should join them into something like `test_boolean_union_intersection`.

**Review by adeak — APPROVED** [review]
Few musing remarks and suggestions, but this could even be merged as-is. I love what you've done to this, now it's completely clear. Thanks.

I've also pushed some changes of my own.

File: pyvista/core/filters/poly_data.py [hunk 3]
```diff
@@ -43,152 +44,292 @@ def edge_mask(poly_data, angle):
...
-        if not isinstance(mesh, pyvista.PolyData):
-            raise TypeError("Input mesh must be PolyData.")
+        # use dataset merge if not polydata
+        if not isinstance(dataset, pyvista.PolyData):
+            if isinstance(dataset, (list, tuple, pyvista.MultiBlock)):
+                for data in dataset:
+                    if not isinstance(data, pyvista.PolyData):
+                        return DataSetFilters.merge(poly_data, dataset,
+                                                    merge_points=merge_points,
+                                                    inplace=inplace)
```
[the remark this answers is not in the document]
Comment (akaszynski) [reply pyvista/core/filters/poly_data.py:307]: 01cf284c245026bd2094ce2f2cbf0932aac7edd6 cleans this up.  Agreed that it's not clean.

File: examples/01-filter/boolean-operations.py [hunk 4]
```diff
@@ -1,110 +1,159 @@
...
+
+Essentially, boolean union, difference, and intersection are all the
+same operation. Just different parts of the objects are kept at the
+end.
 
-For merging, the ``+`` operator can be used between any two meshes in PyVista
-which simply calls the ``.merge()`` filter to combine any two meshes.
-Similarly, the ``-`` operator can be used between any two :class:`pyvista.PolyData`
+The ``-`` operator can be used between any two :class:`pyvista.PolyData`
 meshes in PyVista to cut the first mesh by the second.
```
[the remark this answers is not in the document]
Comment (akaszynski) [reply examples/01-filter/boolean-operations.py:24]: Added in 915a6b74dba954c60adbd3f7177ffdb91a529a0f and 42d5d6dd6.

File: tests/test_polydata.py (hunk 2 above)
Comment (akaszynski) [reply tests/test_polydata.py:288]: done in 9405472c6

File: pyvista/core/filters/poly_data.py [hunk 5]
```diff
@@ -43,152 +44,292 @@ def edge_mask(poly_data, angle):
...
+                                  strips_to_polys=False)
+
+        if inplace:
+            dataset.deep_copy(merged)
+            return dataset
+
+        return merged
 
     def intersection(poly_data, mesh, split_first=True, split_second=True):
         """Compute the intersection between two meshes.
```
[the remark this answers is not in the document]
Comment (akaszynski) [reply pyvista/core/filters/poly_data.py:335]: 43544fc4f

**Comment by akaszynski** [comment]
Thanks for reviewing it @adeak and @MatthewFlamm.  It's a pain, but without you guys keeping this in check, these PRs would be way worse off.

**Pull request merged**